File: src/chipschallenge/tickbehaviors/PinkballTickBehavior.java
package chipschallenge.tickbehaviors;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Move;
import chipschallenge.Move.Moves;

public class PinkballTickBehavior implements BlockTickBehavior {

    private PinkballTickBehavior() {
    }
    private static PinkballTickBehavior mInstance = null;

    public static synchronized PinkballTickBehavior getInstance() {
        if (mInstance == null) {
            mInstance = new PinkballTickBehavior();
        }
        return mInstance;
    }

    @Override
    public void tick(Block caller) throws BlockContainerFullException {
        if(caller.isOnCloner())
            return;
        Moves m = caller.getFacing();
        if (!caller.move(m)) {
            m = Move.reverse(m);
            caller.setFacing(m);
            caller.move(m);
        }
    }
}

File: ChipsChallenge.Shared/Tickbehaviors/PinkballTickBehavior.cs
namespace ChipsChallenge.Shared.Tickbehaviors
{
    using Moves = Move.Moves;

    public class PinkballTickBehavior : IBlockTickBehavior
    {
        private static PinkballTickBehavior instance;

        private PinkballTickBehavior()
        {
        }

        public static PinkballTickBehavior Instance
        {
            get
            {
                lock (typeof(PinkballTickBehavior))
                {
                    return instance ?? (instance = new PinkballTickBehavior());
                }
            }
        }

        public virtual void Tick(Block caller)
        {
            if (caller.OnCloner)
            {
                return;
            }

            Moves moves = caller.Facing;
            if (!caller.Move(moves))
            {
                moves = Move.Reverse(moves);
                caller.Facing = moves;
                caller.Move(moves);
            }
        }
    }
}